test/xbgpu/test_engine.py

```diff
@@ -565,11 +540,23 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
             )
 
         with PromDiff(namespace=METRIC_NAMESPACE) as prom_diff:
-            # Generate one extra chunk to simulate an incomplete accumulation
-            # to check that dumps are aligned correctly - even if the first
-            # received batch is from the middle of an accumulation.
-            batch_start_index = first_batch_index * max_heap_acc_threshold - HEAPS_PER_FENGINE_PER_CHUNK
-            batch_end_index = (first_batch_index + min_full_accumulations) * max_heap_acc_threshold
+            # NOTE: The product of `heap_acc_thresholds` is used in two ways
+            # below. Both uses are to ensure there is a whole number of
+            # accumulations for *both* XPipelines. The first usage is
+            # dual-purpose once more:
+            # - In addition to the above, the arbitrary start position for data
+            #   in this test dictates the first received batch to be in the
+            #   middle of an accumulation.
+            # - More accurately, it forces the first accumulation (to be
+            #   processed) to only contain one batch of data. This ensures
+            #   we test that output dumps are aligned correctly, despite
+            #   the first data processed not being on an accumulation
+            #   boundary.
+            heap_acc_thresholds = [corrprod_output.heap_accumulation_threshold for corrprod_output in corrprod_outputs]
+            first_batch_index = 12 * np.product(heap_acc_thresholds)
+            batch_start_index = first_batch_index - HEAPS_PER_FENGINE_PER_CHUNK
+            batch_end_index = first_batch_index + np.product(heap_acc_thresholds)
```

Comment:
It's a bit confusing to have variables called both `first_batch_index` and `batch_start_index` that are different. Since `first_batch_index` doesn't seem be used again, how about something like this (can also be combined with James's suggestions):
```suggestion
            batch_start_index = 12 * np.product(heap_acc_thresholds)
            batch_end_index = batch_start_index + np.product(heap_acc_thresholds)
            # Add an extra batch before the first full accumulation
            batch_start_index -= HEAPS_PER_FENGINE_PER_CHUNK
```

Reply:
Reading your collective suggestions here, my perceived awkwardness if I'd left `first_batch_index` out is unfounded.

Will update accordingly.